### PR TITLE
DAC6-3039 | Amended Index Page + added subscription connector to registration

### PR DESCRIPTION
**Pull request by matthewharris1**

### CODE REVIEW

**Review by koladeadewuyi — DISMISSED** [review]

**Review by chriswaugh — CHANGES_REQUESTED** [review]

File: app/controllers/IndexController.scala [hunk 1]
```diff
@@ -16,39 +16,51 @@
...
+  subscriptionService: SubscriptionService,
+  conf: FrontendAppConfig,
   view: IndexView
 )(implicit ec: ExecutionContext)
     extends FrontendBaseController
     with Logging
     with I18nSupport {
 
-  def onPageLoad: Action[AnyContent] = identify.async {
+  def onPageLoad(mode: Mode): Action[AnyContent] = identify.async {
```
Comment (chriswaugh) [inline app/controllers/IndexController.scala:45]: I don't think this page needs the concept of a mode. It's always the same content and the same links.

File: app/views/IndexView.scala.html [hunk 2]
```diff
@@ -27,5 +33,15 @@
 
     <h1 class="govuk-heading-xl">@messages("index.heading")</h1>
 
-    <p class="govuk-body">@messages("index.guidance")</p>
+    <p class="govuk-body">@Html(messages("index.p1", link(controllers.addFinancialInstitution.routes.NameOfFinancialInstitutionController.onPageLoad(mode).url, "index.p1.link")))</p>
```
Comment (chriswaugh) [inline app/views/IndexView.scala.html:36]: I'd be happy for us to hardcode the "Normal" mode here and remove the need for having a mode parameter on the route, controller and view for this index page.

File: app/controllers/actions/IdentifierAction.scala [hunk 3]
```diff
@@ -37,41 +40,49 @@ class AuthenticatedIdentifierAction @Inject() (
...
+                                   block: IdentifierRequest[A] => Future[Result]
+  ): Future[Result] = {
 
-  override def invokeBlock[A](request: Request[A], block: IdentifierRequest[A] => Future[Result]): Future[Result] = {
+    val fatcaEnrolment  = "HMRC-FATCA-ORG"
+    val fatcaIdentifier = "FATCAID"
 
-    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+    val subscriptionId: Option[String] = for {
+      enrolment      <- enrolments.getEnrolment(fatcaEnrolment)
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/controllers/actions/IdentifierAction.scala:76]: Amended :)

File: app/controllers/IndexController.scala (hunk 1 above)
Comment (matthewharris1) [reply app/controllers/IndexController.scala]: Removed it

File: app/controllers/actions/IdentifierAction.scala [hunk 4]
```diff
@@ -37,41 +41,49 @@ class AuthenticatedIdentifierAction @Inject() (
...
-    extends IdentifierAction {
+  private def getSubscriptionId[A](request: Request[A],
+                                   enrolments: Enrolments,
+                                   internalId: String,
+                                   affinityGroup: AffinityGroup,
+                                   block: IdentifierRequest[A] => Future[Result]
+  ): Future[Result] = {
 
-  override def invokeBlock[A](request: Request[A], block: IdentifierRequest[A] => Future[Result]): Future[Result] = {
+    val fatcaEnrolment  = config.enrolmentKey
```
Comment (chriswaugh) [inline app/controllers/actions/IdentifierAction.scala:72]: Looks like these two variables are only used in the for comprehension below so I don't think they need to be variables and you could just call the config/object directly in the comprehension? e.g.:

```
  enrolment <- enrolments.getEnrolment(config.enrolmentKey)
  id <- enrolment.getIdentifier(IdentifierType.FATCAID)
```

File: app/controllers/auth/AuthController.scala [hunk 5]
```diff
@@ -41,7 +41,7 @@ class AuthController @Inject() (
         .clear(request.userId)
         .map {
           _ =>
-            Redirect(config.signOutUrl, Map("continue" -> Seq(config.exitSurveyUrl)))
+            Redirect(config.signOutUrl).withSession()
```
[the remark this answers is not in the document]
Comment (chriswaugh) [reply app/controllers/auth/AuthController.scala:44]: Can we have a look at this together this afternoon? Just want to make sure I'm understanding it right as this function is one of the ones that comes out of the box and I know when we've tweaked it in other services before it's broken things. Just want to understand what tests were breaking & what our requirements are.

**Review by chriswaugh — CHANGES_REQUESTED** [review]
Looking good! Only one small suggestion but I'm probably being picky.

Main thing is the changes to the AuthController, I'd like to understand a bit more so can we run through it together when you're next free?

File: app/controllers/IndexController.scala [hunk 6]
```diff
@@ -16,39 +16,51 @@
 
 package controllers
 
+import config.FrontendAppConfig
 import controllers.actions.IdentifierAction
-import models.UserAnswers
+import models.{Mode, UserAnswers}
```
Comment (chriswaugh) [inline app/controllers/IndexController.scala:21]: This import is unused

**Review by chriswaugh — APPROVED** [review]

**Pull request merged**